volume/store/store.go

```diff
@@ -47,39 +60,94 @@ func (s *VolumeStore) AddAll(vols []volume.Volume) {
 func (s *VolumeStore) Create(name, driverName string, opts map[string]string) (volume.Volume, error) {
 	s.mu.Lock()
 	if vc, exists := s.vols[name]; exists {
-		v := vc.Volume
 		s.mu.Unlock()
-		return v, nil
+		<-vc.waitCreate
+		if vc.Volume != nil && (vc.Volume.DriverName() == driverName || driverName == volume.DefaultDriverName) {
+			return vc.Volume, nil
+		}
+		return nil, ErrNameConflict
 	}
+
+	vc := newVolumeCounter()
+	s.vols[name] = vc
 	s.mu.Unlock()
 	logrus.Debugf("Registering new volume reference: driver %s, name %s", driverName, name)
 
-	vd, err := volumedrivers.GetDriver(driverName)
+	var (
+		err error
+		vd  volume.Driver
+		v   volume.Volume
+	)
+
+	defer func() {
+		if err != nil {
+			s.mu.Lock()
+			delete(s.vols, name)
+			s.mu.Unlock()
+		}
+		close(vc.waitCreate)
```

Comment:
plugin 2 would not be holding the mu.Lock as it unlocks before waiting for vc.waitCreate.